File: core/MergeKSortedLists/MergeKSortedLists.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Queue;

public class MergeKSortedLists {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        Queue<int[]> queue = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        Map<Integer, ListNode> map = new HashMap<>();
        int count = 0;
        for (ListNode node : lists) {
            if (node != null) {
                queue.offer(new int[]{node.val, count});
                map.put(count, node);
                count++;
            }
        }

        ListNode res = new ListNode();
        ListNode preHead = res;
        while (!queue.isEmpty()) {
            int[] nodeCount = queue.poll();
            ListNode node = map.get(nodeCount[1]);
            preHead.next = node;
            preHead = preHead.next;
            if (node.next != null) {
                queue.offer(new int[]{node.next.val, count});
                map.put(count, node.next);
                count++;
            }
        }
        return res.next;
    }
}

File: core/MergeKSortedLists/MergeKSortedLists.py
from typing import List, Optional
import heapq


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def mergeKLists(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    queue, remain, count = [], {}, 0
    for node in lists:
        if node:
            remain[count] = node
            heapq.heappush(queue, (node.val, count))
            count += 1
    pre_head = ListNode()
    res = pre_head
    while queue:
        _, node_count = heapq.heappop(queue)
        pre_head.next = remain[node_count]
        pre_head = pre_head.next
        if remain[node_count].next:
            remain[count] = remain[node_count].next
            heapq.heappush(queue, (remain[node_count].next.val, count))
            count += 1

    return res.next
